web/pandas/pdeps/0006-ban-upcasting.md

```diff
@@ -50,25 +50,28 @@ In [10]: ser[2] = '2000-01-04'  # works, is converted to datetime64
 In [11]: ser[2] = '2000-01-04x'  # almost certainly a typo - but pandas doesn't error, it upcasts to object
 ```
 
-The scope of this PDEP is limited to setitem-like operations which would operate inplace, such as:
-- ``ser[0] = 2.5``;
+The scope of this PDEP is limited to setitem-like operations on Series.
+For example, starting with
+```python
+df = DataFrame({'a': [1, 2, np.nan], 'b': [4, 5, 6]})
+ser = df['a'].copy()
+```
+then the following would all raise:
+- ``ser[0] = 'foo'``;
 - ``ser.fillna('foo', inplace=True)``;
 - ``ser.where(ser.isna(), 'foo', inplace=True)``
-- ``ser.iloc[0] = 2.5``
-- ``ser.loc[0] = 2.5``
-- ``ser[:] = 2.5``
+- ``ser.iloc[0] = 'foo'``
+- ``ser.loc[0] = 'foo'``
+- ``df.loc[0, 'a'] = 'foo'``
 
-There may be more. What is explicitly excluded from this PDEP is any operation would have no change
-of operating inplace to begin with, such as:
+Examples of operations which would not raise are:
 - ``ser.diff()``;
-- ``pd.concat([ser, other])``;
+- ``pd.concat([ser, ser.astype(object)])``;
 - ``ser.mean()``;
-- ``df.loc[0, 'col1'] = 2.5`` (if ``df`` is not a single block).
+- ``df.loc[:, 'a'] = 'foo'`` (debatable, as is the one below)
+- ``ser[:] = 'foo'``
```

Comment:
What about `ser.iloc[[True, False, True]] = 'foo'` ?

If something like `ser.where()` would raise, because it is changing multiple elements, why wouldn't any kind of slice also raise?

Reply:
I think that comment stems from the time before https://github.com/pandas-dev/pandas/pull/45333 when we still special cased "full" slices (`:`) to overwrite data, instead of updating inplace. 

Now that any slice (whether it is a subset or "accidentally" slicing all data) always first tries to set inplace, I think for this PDEP we also shouldn't special case this anymore.

That would mean that `ser[:] = ..` doesn't do anything else compared to eg `ser[1:3]`

Reply:
Nice, thanks - I like that, avoids value-dependent behaviour if possible